File: golang/src/LSP/problem/pricing_utils.go
package problem_lsp

type PricingUtils struct {
}

func Run() {
	p1 := NewProduct()
	p2 := NewProduct()
	//p3 := NewInHouseProduct()

	productList := make([]*Product,3)

	productList = append(productList,p1)
	productList = append(productList,p2)
	//productList = append(productList,p3)

	//for (Product product: productList) {
	//
	//	if(product instanceof InHouseProduct) {
	//		((InHouseProduct) product).applyExtraDiscount();
	//	}
	//	System.out.println(product.getDiscount());

}

File: javalang/src/LSP/problem/PricingUtils.java
package LSP.problem;

import java.util.ArrayList;
import java.util.List;

public class PricingUtils {
	
	public static void main(String[] args) {

		Product p1 = new Product();
		Product p2 = new Product();
		Product p3 = new InHouseProduct();
	
		List<Product> productList = new ArrayList<>();
		
		productList.add(p1);
		productList.add(p2);
		productList.add(p3);
		
		for (Product product: productList) {
			
			if(product instanceof InHouseProduct) {
				((InHouseProduct) product).applyExtraDiscount();
			}
			System.out.println(product.getDiscount());
			
		}
	}
	
}
